src/dune_pkg/fetch.ml

```diff
@@ -188,18 +188,78 @@ type failure =
 
 let label = "dune-fetch"
 
+module Single_run_cache = struct
+  type entry =
+    { filename : Filename.t
+    ; download_complete : unit Fiber.t
+    }
+
+  (* A cache of downloaded files keyed by urls which persists within a single
+     invocation of dune and is deleted when dune terminates *)
+  type t =
+    { temp_dir : Path.t
+    ; url_to_filename : entry String.Map.t ref
+    }
+
+  let create () =
+    let temp_dir = Temp.create Dir ~prefix:"dune" ~suffix:"fetch-archive-cache" in
+    at_exit (fun () -> Temp.destroy Dir temp_dir);
+    { temp_dir; url_to_filename = ref String.Map.empty }
```

Comment:
If this is a `Map ref`, maybe `Hashtbl` could be an option. It has slightly different semantics.

Reply:
done